internal/core/docker_context.go

```diff
@@ -0,0 +1,328 @@
+	// DockerEndpoint is the name of the docker endpoint in a stored context
+	dockerEndpoint string = "docker"
+)
+
+// dockerContext is a typed representation of what we put in Context metadata
+type dockerContext struct {
+	Description      string
+	AdditionalFields map[string]any
+}
+
+type metadataStore struct {
+	root   string
+	config contextConfig
+}
+
+// typeGetter is a func used to determine the concrete type of a context or
+// endpoint metadata by returning a pointer to an instance of the object
+// eg: for a context of type DockerContext, the corresponding typeGetter should return new(DockerContext)
+type typeGetter func() any
+
+// namedTypeGetter is a typeGetter associated with a name
+type namedTypeGetter struct {
+	name       string
+	typeGetter typeGetter
+}
+
+// endpointTypeGetter returns a namedTypeGetter with the specified name and getter
+func endpointTypeGetter(name string, getter typeGetter) namedTypeGetter {
+	return namedTypeGetter{
+		name:       name,
+		typeGetter: getter,
+	}
+}
+
+// endpointMeta contains fields we expect to be common for most context endpoints
+type endpointMeta struct {
+	Host          string `json:",omitempty"`
+	SkipTLSVerify bool
+}
+
+var defaultStoreEndpoints = []namedTypeGetter{
+	endpointTypeGetter(dockerEndpoint, func() any { return &endpointMeta{} }),
+}
+
+// contextConfig is used to configure the metadata marshaler of the context ContextStore
+type contextConfig struct {
+	contextType   typeGetter
+	endpointTypes map[string]typeGetter
+}
+
+// newConfig creates a config object
+func newConfig(contextType typeGetter, endpoints ...namedTypeGetter) contextConfig {
+	res := contextConfig{
+		contextType:   contextType,
+		endpointTypes: make(map[string]typeGetter),
+	}
+
+	for _, e := range endpoints {
+		res.endpointTypes[e.name] = e.typeGetter
+	}
+	return res
+}
+
+// metadata contains metadata about a context and its endpoints
+type metadata struct {
+	Name      string         `json:",omitempty"`
+	Metadata  any            `json:",omitempty"`
+	Endpoints map[string]any `json:",omitempty"`
+}
+
+func (s *metadataStore) contextDir(id contextdir) string {
+	return filepath.Join(s.root, string(id))
+}
+
+type untypedContextMetadata struct {
+	Metadata  json.RawMessage            `json:"metadata,omitempty"`
+	Endpoints map[string]json.RawMessage `json:"endpoints,omitempty"`
+	Name      string                     `json:"name,omitempty"`
+}
+
+func (s *metadataStore) getByID(id contextdir) (metadata, error) {
+	fileName := filepath.Join(s.contextDir(id), metaFile)
+	bytes, err := os.ReadFile(fileName)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return metadata{}, errdefs.NotFound(fmt.Errorf("context not found: %w", err))
+		}
+		return metadata{}, err
+	}
+
+	var untyped untypedContextMetadata
+	r := metadata{
+		Endpoints: make(map[string]any),
+	}
+
+	if err := json.Unmarshal(bytes, &untyped); err != nil {
+		return metadata{}, fmt.Errorf("parsing %s: %w", fileName, err)
```

Comment:
suggestion: use a unique wrap between this and the other errors below so we can identify where the failure occurred.